libexec/rtld-elf/rtld_c18n.c

```diff
@@ -0,0 +1,770 @@
+		}
+		atomic_flag_clear_explicit(&tramp_pgs.lock,
+		    memory_order_release);
+	}
+	assert(cheri_gettag(tramp));
+
+	entry = tramp + (entry - tmp);
+	memcpy(tramp, tmp, len);
+
+	/*
+	 * Ensure i- and d-cache coherency after writing executable code. The
+	 * __clear_cache procedure rounds the addresses to cache-line-aligned
+	 * addresses. Derive the start/end addresses from pg so that they have
+	 * sufficiently large bounds to contain these rounded addresses.
+	 */
+	__clear_cache(cheri_copyaddress(pg, entry),
+	    cheri_copyaddress(pg, tramp + len));
+
+	entry = cheri_clearperm(entry, FUNC_PTR_REMOVE_PERMS);
+	return (cheri_sealentry(cheri_capmode(entry)));
+}
+
+static bool
+tramp_sig_equal(struct tramp_sig lhs, struct tramp_sig rhs)
+{
+	return (lhs.reg_args == rhs.reg_args &&
+		lhs.mem_args == rhs.mem_args &&
+		lhs.ret_args == rhs.ret_args);
+}
+
+static bool
+tramp_sig_compatible(struct tramp_sig lhs, struct tramp_sig rhs)
+{
+	return (!lhs.valid || !rhs.valid || tramp_sig_equal(lhs, rhs));
+}
+
+static void *
+tramp_get_entry(const struct tramp_data *found, const struct tramp_data *data)
+{
+	if (!tramp_sig_compatible(found->sig, data->sig))
+		rtld_fatal(
+		    "Incompatible signatures for function %s: "
+		    "%s requests %02hhX but %s provides %02hhX",
+		    strtab_value(data->defobj, data->def->st_name),
+		    data->defobj->path, tramp_sig_to_int(data->sig),
+		    found->defobj->path, tramp_sig_to_int(found->sig));
+
+	return (found->entry);
+}
+
+static void *
+tramp_create_entry(struct tramp_data *found, const struct tramp_data *data)
+{
+	struct tramp_sig sig;
+
+	*found = *data;
+	if (found->def != NULL) {
+		sig = tramp_fetch_sig(found->defobj,
+		    found->def - found->defobj->symtab);
+		if (!found->sig.valid)
+			found->sig = sig;
+		else if (!tramp_sig_compatible(found->sig, sig))
+			rtld_fatal(
+			    "Incompatible signatures for function %s: "
+			    "requests %02hhX but %s provides %02hhX",
+			    strtab_value(found->defobj, found->def->st_name),
+			    tramp_sig_to_int(found->sig), found->defobj->path,
+			    tramp_sig_to_int(sig));
+	}
+
+	return (found->entry = tramp_pgs_append(found));
+}
+
+void *
+tramp_intern(const Obj_Entry *reqobj, const struct tramp_data *data)
+{
+	void *entry;
+	RtldLockState lockstate;
+	ptraddr_t target = (ptraddr_t)data->target;
+	const uint32_t hash = pointer_hash(target);
+	slot_idx_t slot, idx;
+	ptraddr_t key;
+	int exp;
+
+	/* reqobj == NULL iff the request is by RTLD */
+	assert(
+	    (reqobj == NULL || data->def != NULL) &&
+	    data->defobj != NULL &&
+	    data->entry == NULL &&
+	    tramp_sig_legal(data->sig));
+
+	if (!tramp_should_include(reqobj, data))
+		return (data->target);
+
+start:
+	slot = hash;
+	rlock_acquire(rtld_tramp_lock, &lockstate);
+	exp = tramp_table.exp;
+	do {
+		slot = nextSlot(slot, hash, exp);
+		key = atomic_load_explicit(&tramp_table.map[slot].key,
+		    memory_order_relaxed);
+		if (key != 0)
+			continue;
+		/*
```

Comment:
Throughout this function: multi-line comments need a blank line before them, and loops need a blank line after them